add multiple sensors observing feature

catalinghita8 opened the pull request:
Improvement that fixes #33.

Added the possibility of observing and getting data flow from multiple sensors.

catalinghita8 committed:
add multiple sensors observing feature

pwittchen commented:
Hi @catalinghita8,

Thanks for your PR and your effort. Unfortunately, you reformatted the whole file, so I cannot see your exact changes within PR without analyzing the whole file. Please, use code style described here: https://github.com/pwittchen/ReactiveSensors#code-style. Moreover, I don't see any unit tests related to the feature you introduced.

Regards,
Piotr

pwittchen reviewed (changes requested):
Hi @catalinghita8,

Thanks for your PR and your effort. Unfortunately, you reformatted the whole file with different code style than the one, which is used in the project, so I cannot see your exact changes within PR without analyzing the whole file. Please, use code style described here: https://github.com/pwittchen/ReactiveSensors#code-style. Moreover, I don't see any unit tests related to the feature you introduced.

Regards,
Piotr

catalinghita8 commented:
Hi @pwittchen,
I completely missed the coding style - my bad on that - thus adjusted the changes properly now so you can review them.

Regarding adding unit tests, I am aware that I haven't added any, mainly because currently there are no unit tests (beside the basic one) and I don't want to add new dependencies (like Mockito) into the project. It usually requires some responsability and I believe that the creator should decide upon adding new dependencies as it might have repercussions on the library.

Also you might decide enforcing some style for the tests as well. Therefore, once you review the code and decide on including some basic tests, I will be more than glad to write some - following your pattern  - for this feature as well.

All the best,
Catalin

pwittchen commented:
Hi,

Thanks for update. Can you squash your commits into one? It'll be easier to view changes in the PR and git history will be organized better. You can use this tool: http://rebaseandsqua.sh/.

Regarding unit tests, you're right. This project unfortunately has poor tests. I usually add them in my projects, but I didn't have more time to work on that in this case, so we can skip tests in your PR and test infrastructure/configuration can be improved later.

catalinghita8 commented:
Thanks for the feedback. Just rebased and squashed that redundant commit. Now it should be easier to read.

Catalin

pwittchen commented on library/src/main/java/com/github/pwittchen/reactivesensors/library/ReactiveSensors.java:
```diff
@@ -92,11 +93,30 @@ public boolean hasSensor(int sensorType) {
+   * @param sensorTypes sensor type from Sensor class from Android SDK
+   * @return RxJava Observable with ReactiveSensorEvent
+   */
+  public Flowable<ReactiveSensorEvent> observeSensors(final int samplingPeriodInUs,
```
I'd rename this method to e.g. `observeManySensors` - `observeSensor` and `observeSensors` looks similar and it's easy to make a mistake while using such API.

pwittchen commented on library/src/main/java/com/github/pwittchen/reactivesensors/library/ReactiveSensors.java:
```diff
@@ -134,7 +155,72 @@ public boolean hasSensor(int sensorType) {
+    StringBuilder errorMessage = null;
+    for (int sensorType : sensorTypes) {
+      if (!hasSensor(sensorType)) {
+        if (errorMessage == null) {
```
I think, it can be simplified like that:

```java
if(errorMessage == null) {
  errorMessage = new StringBuilder("Following sensors are not available on current device: ");
}

errorMessage.append(sensorType);
errorMessage.append(" ");
```

In such case, `else` block could be skipped.

pwittchen commented on library/src/main/java/com/github/pwittchen/reactivesensors/library/ReactiveSensors.java:
```diff
@@ -134,7 +155,72 @@ public boolean hasSensor(int sensorType) {
+   * @param sensorTypes sensor types array from Sensor class from Android SDK
+   * @return RxJava Observable with ReactiveSensorEvent
+   */
+  public Flowable<ReactiveSensorEvent> observeSensors(final int samplingPeriodInUs,
```
This method has around ~50 lines of code. It's too many. Please, try to split it into smaller methods, so the max length of the single method will have 20 lines of code or less.

pwittchen commented:
Thanks for that. I had a few issues in the code review. Please have a look at them.

catalinghita8 committed:
refactor multiple reactive sensors feature

catalinghita8 commented:
@pwittchen apologies for the late response and cheers for the constructive suggestions! 

Refactored the multiple sensor observing method naming as well as the the method's content in cause.
Extracted the error message creation outside of the main method, thus reducing the size from ~50 to ~30 lines. 

Extra:
As you requested the method to be under ~20, I will have to insist in keeping it to the actual size. Any further trimming by extracting essential pieces of the code that make sense in the method's context, would result in external methods that have no modular actions or even meaning.
While I agree methods should not be large and should be kept as minimal as possible,  please note that the `observeSensor` has ~30 lines itself, thus I see no valid points in reducing the `observeMultipleSensors` to under 30 lines.

pwittchen commented:
Thanks, I'll have a look at it in one week from now. I'm away from home
without my laptop now. :-)

--
Piotr Wittchen,
http://wittchen.io

sob., 27 kwi 2019, 15:27 użytkownik catalinghita8 <notifications@github.com>
napisał:

> @pwittchen <https://github.com/pwittchen> apologies for the late response
> and cheers for the constructive suggestions!
>
> Refactored the multiple sensor observing method naming as well as the the
> method's content in cause.
> Extracted the error message creation outside of the main method, thus
> reducing the size from ~50 to ~30 lines.
>
> Extra:
> As you requested the method to be under ~20, I will have to insist in
> keeping it to the actual size. Any further trimming by extracting essential
> pieces of the code that make sense in the method's context, would result in
> external methods that have no modular actions or even meaning.
> While I agree methods should not be large and should be kept as minimal as
> possible, please note that the observeSensor has ~30 lines itself, thus I
> see no valid points in reducing the observeMultipleSensors to under 30
> lines.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pwittchen/ReactiveSensors/pull/37#issuecomment-487290496>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFJYFZKY32RBXMSOG52JDDPSRPFRANCNFSM4GX6QKJQ>
> .
>

catalinghita8 commented:
> Thanks, I'll have a look at it in one week from now. I'm away from home without my laptop now. :-)
> […](#)
> -- Piotr Wittchen, http://wittchen.io sob., 27 kwi 2019, 15:27 użytkownik catalinghita8 <notifications@github.com> napisał:
> @pwittchen <https://github.com/pwittchen> apologies for the late response and cheers for the constructive suggestions! Refactored the multiple sensor observing method naming as well as the the method's content in cause. Extracted the error message creation outside of the main method, thus reducing the size from ~50 to ~30 lines. Extra: As you requested the method to be under ~20, I will have to insist in keeping it to the actual size. Any further trimming by extracting essential pieces of the code that make sense in the method's context, would result in external methods that have no modular actions or even meaning. While I agree methods should not be large and should be kept as minimal as possible, please note that the observeSensor has ~30 lines itself, thus I see no valid points in reducing the observeMultipleSensors to under 30 lines. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#37 (comment)](https://github.com/pwittchen/ReactiveSensors/pull/37#issuecomment-487290496)>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAFJYFZKY32RBXMSOG52JDDPSRPFRANCNFSM4GX6QKJQ> .

@pwittchen Let me know if anything is needed.

pwittchen commented:
Ok

--
Piotr Wittchen,
http://wittchen.io

pt., 28 cze 2019, 17:16 użytkownik catalinghita8 <notifications@github.com>
napisał:

> Thanks, I'll have a look at it in one week from now. I'm away from home
> without my laptop now. :-)
> … <#m_8316246545055525711_>
> -- Piotr Wittchen, http://wittchen.io sob., 27 kwi 2019, 15:27 użytkownik
> catalinghita8 notifications@github.com napisał:
> @pwittchen <https://github.com/pwittchen> https://github.com/pwittchen
> apologies for the late response and cheers for the constructive
> suggestions! Refactored the multiple sensor observing method naming as well
> as the the method's content in cause. Extracted the error message creation
> outside of the main method, thus reducing the size from ~50 to ~30 lines.
> Extra: As you requested the method to be under ~20, I will have to insist
> in keeping it to the actual size. Any further trimming by extracting
> essential pieces of the code that make sense in the method's context, would
> result in external methods that have no modular actions or even meaning.
> While I agree methods should not be large and should be kept as minimal as
> possible, please note that the observeSensor has ~30 lines itself, thus I
> see no valid points in reducing the observeMultipleSensors to under 30
> lines. — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub <#37 (comment)
> <https://github.com/pwittchen/ReactiveSensors/pull/37#issuecomment-487290496>>,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAFJYFZKY32RBXMSOG52JDDPSRPFRANCNFSM4GX6QKJQ
> .
>
> @pwittchen <https://github.com/pwittchen> Let me know if anything is
> needed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pwittchen/ReactiveSensors/pull/37?email_source=notifications&email_token=AAFJYFZPEBYFVDKMVELE54LP4YTMHA5CNFSM4GX6QKJ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY2LQVQ#issuecomment-506771542>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFJYF6WTGNUB7HCOPF7BOLP4YTMHANCNFSM4GX6QKJQ>
> .
>

catalinghita8 commented:
@pwittchen As a follow-up, the changes you requested have been applied, I am still waiting for your actions.

pwittchen commented:
Sorry, I've been quite busy, so I didn't take care of it. I'll try to push it forward soon.

pwittchen commented:
Changes merged. They'll be available in the next release.

I haven't released new versions of my libraries from some time. 
I'll try to release them quite soon due to new updates and fresh dependencies.

catalinghita8 commented:
@pwittchen thanks!
